core/src/Kokkos_NumericTraits.hpp

```diff
@@ -45,347 +45,61 @@
-    return static_cast<long long>(0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static long long land() {
-    return static_cast<long long>(1);
-  }
-};
-
-template <>
-struct reduction_identity<unsigned char> {
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned char sum() {
-    return static_cast<unsigned char>(0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned char prod() {
-    return static_cast<unsigned char>(1);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned char max() {
-    return static_cast<unsigned char>(0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned char min() {
-    return UCHAR_MAX;
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned char bor() {
-    return static_cast<unsigned char>(0x0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned char band() {
-    return ~static_cast<unsigned char>(0x0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned char lor() {
-    return static_cast<unsigned char>(0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned char land() {
-    return static_cast<unsigned char>(1);
-  }
-};
 
-template <>
-struct reduction_identity<unsigned short> {
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned short sum() {
-    return static_cast<unsigned short>(0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned short prod() {
-    return static_cast<unsigned short>(1);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned short max() {
-    return static_cast<unsigned short>(0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned short min() {
-    return USHRT_MAX;
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned short bor() {
-    return static_cast<unsigned short>(0x0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned short band() {
-    return ~static_cast<unsigned short>(0x0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned short lor() {
-    return static_cast<unsigned short>(0);
+  KOKKOS_FORCEINLINE_FUNCTION constexpr static T prod() {
+    return static_cast<T>(1);
   }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned short land() {
-    return static_cast<unsigned short>(1);
-  }
-};
 
-template <>
-struct reduction_identity<unsigned int> {
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned int sum() {
-    return static_cast<unsigned int>(0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned int prod() {
-    return static_cast<unsigned int>(1);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned int max() {
-    return static_cast<unsigned int>(0);
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned int min() {
-    return UINT_MAX;
-  }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static unsigned int bor() {
-    return static_cast<unsigned int>(0x0);
+  KOKKOS_FORCEINLINE_FUNCTION constexpr static T max() {
+    return std::numeric_limits<T>::lowest();
```

Comment:
Is it marked `__host__ __device__`??

Reply:
Yeah, we likely need `KOKKOS_ENABLE_CUDA_CONSTEXPR` for this to work reliably.

Reply:
Let's just fix the typo. See #2802